### PR TITLE
feat(release): allow specifying npm dist-tag per branch

### PR DESCRIPTION
**Pull request by eladb**
Adds `npmDistTag` to release branch definition which will be configured in the release workflow triggered for this branch.

Implementation notes: to support this we needed to refactor the publisher so that when its jobs are rendered, they have the branch context. This also means that now we can ensire that when a publish command is executed manually, it will be done only on the correct release branch.

The root `npmDistTag` setting applies to the default branch.

BREAKING CHANGE: `npmDistTag` only effects publishing (and supported for each release branch) setting and so it is no longer set in `package.json`.
* `workflow.addJobsLater()` is no longer supported. Use `postSynthesis` hooks and call `.addJobs()` as needed.
* `publish:xxxx` tasks for automated releases are not defined by default since normally they should only be executed from within workflows. Set `publishingTasks: true` to create them.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by eladb** [comment]
Copy: @njlynch -- this is for `constructs`.

File: test/release/release.test.ts [hunk 1]
```diff
@@ -500,3 +522,117 @@ test('publisher can use custom github runner', () => {
...
+describe('npmDistTag', () => {
+  test('determines npm dist-tag used in the workflow', () => {
+    // GIVEN
+    const project = new TestProject();
+
+    // WHEN
+    const release = new Release(project, {
+      releaseBranches: {
+        'main-3': { majorVersion: 3, npmDistTag: 'latest-3' },
+      },
```
Comment (Chriscbr) [inline test/release/release.test.ts:535]: I assume the outputs are the same if the branch is specified through `addBranch()`? I think this is probably caught by other test cases but just double checking

**Review by Chriscbr — APPROVED** [review]
Overall strategy looks good to me. I wasn't a big fan of `addJobsLater()` either 🙃

File: src/release/publisher.ts [hunk 2]
```diff
@@ -315,91 +352,104 @@ export class Publisher extends Component implements IJobProvider {
...
-    });
+    }));
   }
 
-  private addPublishJob(opts: PublishJobOptions) {
-    const jobname = `release_${opts.name}`;
-    if (jobname in this._jobs) {
-      throw new Error(`Duplicate job with name "${jobname}"`);
-    }
+  private addPublishJob(factory: (branch: string, branchOptions: Partial<BranchOptions>) => PublishJobOptions) {
```
[the remark this answers is not in the document]
Comment (eladb) [reply src/release/publisher.ts:358]: Meh... As long as it's internal

File: test/release/release.test.ts (hunk 1 above)
Comment (eladb) [reply test/release/release.test.ts:535]: Everything going through `addBranch()` eventually so it's covered.

Comment (eladb) [reply test/release/release.test.ts:535]: I would be nice if the main branch and the "other" branches are all defined under `releaseBranches` but I got a bit tangled with the defaults... I'll revisit sometime.

**Pull request merged**